services/brig/src/Brig/Data/MLS/KeyPackage.hs

```diff
@@ -97,6 +102,34 @@ derefKeyPackage ref = do
     q :: PrepQuery R (Identity KeyPackageRef) (Domain, UserId, ClientId)
     q = "SELECT domain, user, client from mls_key_package_refs WHERE ref = ?"
 
+keyPackageRefConvId :: MonadClient m => KeyPackageRef -> MaybeT m (Qualified ConvId)
+keyPackageRefConvId ref = MaybeT $ do
+  qr <- retry x1 $ query1 q (params LocalSerial (Identity ref))
+  pure $ do
+    (domain, cid) <- qr
+    Qualified <$> cid <*> domain
+  where
+    q :: PrepQuery R (Identity KeyPackageRef) (Maybe Domain, Maybe ConvId)
+    q = "SELECT conv_domain, conv FROM mls_key_package_refs WHERE ref = ?"
+
+-- We want to proper update, not an upsert, to avoid "ghost" refs without user+client
+keyPackageRefSetConvId :: MonadClient m => KeyPackageRef -> Qualified ConvId -> m Bool
```

Comment:
This is probably ok, but we might also want to consider the more simple-minded approach of checking existence first, then updating, as we do in various other places.